### ISSUE TITLE
ICE with aligned enum

### ISSUE THREAD

**Issue by MiSawa** [rust-lang/rust#92464]
### Code

```Rust
#[allow(dead_code)]
#[repr(align(8))]
enum Aligned {
    Zero = 0,
    One = 1,
}

fn main() {
    let aligned = Aligned::Zero;
    println!("{}", aligned as u8);
}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.59.0-nightly (b60e32c82 2021-12-30)
binary: rustc
commit-hash: b60e32c82864d841e87359333af1e6d1f9cff9ee
commit-date: 2021-12-30
host: x86_64-unknown-linux-gnu
release: 1.59.0-nightly
LLVM version: 13.0.0
```

### Error output

```
error: internal compiler error: /rustc/b60e32c82864d841e87359333af1e6d1f9cff9ee/compiler/rustc_const_eval/src/interpret/operand.rs:324:13: primitive read failed for type: Aligned
  --> ./ice.rs:10:14
   |
10 |     println!("{}", aligned as u8);
   |              ^^^^

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/b60e32c82864d841e87359333af1e6d1f9cff9ee/compiler/rustc_errors/src/lib.rs:1115:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.59.0-nightly (b60e32c82 2021-12-30) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [optimized_mir] optimizing MIR for `main`
#1 [collect_and_partition_mono_items] collect_and_partition_mono_items
end of query stack
error: aborting due to previous error
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
error: internal compiler error: /rustc/b60e32c82864d841e87359333af1e6d1f9cff9ee/compiler/rustc_const_eval/src/interpret/operand.rs:324:13: primitive read failed for type: Aligned
  --> ./ice.rs:10:14
   |
10 |     println!("{}", aligned as u8);
   |              ^^^^

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/b60e32c82864d841e87359333af1e6d1f9cff9ee/compiler/rustc_errors/src/lib.rs:1115:9
stack backtrace:
   0: std::panicking::begin_panic::<rustc_errors::ExplicitBug>
   1: std::panic::panic_any::<rustc_errors::ExplicitBug>
   2: <rustc_errors::HandlerInner>::span_bug::<rustc_span::span_encoding::Span>
   3: <rustc_errors::Handler>::span_bug::<rustc_span::span_encoding::Span>
   4: rustc_middle::ty::context::tls::with_opt::<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, ()>
   5: rustc_middle::util::bug::opt_span_bug_fmt::<rustc_span::span_encoding::Span>
   6: rustc_middle::util::bug::span_bug_fmt::<rustc_span::span_encoding::Span>
   7: <rustc_const_eval::interpret::eval_context::InterpCx<rustc_mir_transform::const_prop::ConstPropMachine>>::eval_rvalue_into_place
   8: <rustc_mir_transform::const_prop::ConstPropagator as rustc_middle::mir::visit::MutVisitor>::visit_body
   9: <rustc_mir_transform::const_prop::ConstProp as rustc_middle::mir::MirPass>::run_pass
  10: rustc_mir_transform::pass_manager::run_passes
  11: rustc_mir_transform::optimized_mir
  12: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<rustc_span::def_id::DefId, &rustc_middle::mir::Body>>
  13: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::optimized_mir
  14: <rustc_middle::ty::context::TyCtxt>::instance_mir
  15: rustc_monomorphize::collector::collect_neighbours
  16: rustc_monomorphize::collector::collect_items_rec
  17: <rustc_session::session::Session>::time::<(), rustc_monomorphize::collector::collect_crate_mono_items::{closure#1}>
  18: rustc_monomorphize::collector::collect_crate_mono_items
  19: rustc_monomorphize::partitioning::collect_and_partition_mono_items
  20: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<(), (&std::collections::hash::set::HashSet<rustc_span::def_id::DefId, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>, &[rustc_middle::mir::mono::CodegenUnit])>>
  21: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::collect_and_partition_mono_items, rustc_query_impl::plumbing::QueryCtxt>
  22: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::collect_and_partition_mono_items
  23: <rustc_codegen_llvm::LlvmCodegenBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::codegen_crate
  24: <rustc_session::session::Session>::time::<alloc::boxed::Box<dyn core::any::Any>, rustc_interface::passes::start_codegen::{closure#0}>
  25: <rustc_interface::queries::Queries>::ongoing_codegen
  26: <rustc_interface::interface::Compiler>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}, core::result::Result<core::option::Option<rustc_interface::queries::Linker>, rustc_errors::ErrorReported>>
  27: rustc_span::with_source_map::<core::result::Result<(), rustc_errors::ErrorReported>, rustc_interface::interface::create_compiler_and_run<core::result::Result<(), rustc_errors::ErrorReported>, rustc_driver::run_compiler::{closure#1}>::{closure#1}>
  28: rustc_interface::interface::create_compiler_and_run::<core::result::Result<(), rustc_errors::ErrorReported>, rustc_driver::run_compiler::{closure#1}>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.59.0-nightly (b60e32c82 2021-12-30) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [optimized_mir] optimizing MIR for `main`
#1 [collect_and_partition_mono_items] collect_and_partition_mono_items
end of query stack
error: aborting due to previous error
```

</p>
</details>

**Comment by MiSawa** [comment]
FYI, also fails in stable `rustc 1.57.0 (f1edd0429 2021-11-29)`. https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=72d716ba5a7f274d8c518b82236688b8

**Comment by bugadani** [comment]
Slightly different example, different error message, same cause:

```rust
#[repr(align(2))] // 2, 4, 8 crashes, 1 compiles
enum Aligned {
    Zero = 0,
    One = 1,
}

pub fn main() {
    println!("{}", tou8(Aligned::Zero))
}

fn tou8(al: Aligned) -> u8 {
    al as u8
}
```

`
error: internal compiler error: /rustc/b60e32c82864d841e87359333af1e6d1f9cff9ee/compiler/rustc_codegen_ssa/src/mir/operand.rs:118:18: not immediate: OperandRef(Ref((%Aligned*:  %2 = alloca %Aligned, align 2), None, Align { pow2: 1 }) @ TyAndLayout { ty: Aligned, layout: Layout { fields: Arbitrary { offsets: [Size { raw: 0 }], memory_index: [0] }, variants: Multiple { tag: Scalar { value: Int(I8, false), valid_range: 0..=1 }, tag_encoding: Direct, tag_field: 0, variants: [Layout { fields: Arbitrary { offsets: [], memory_index: [] }, variants: Single { index: 0 }, abi: Aggregate { sized: true }, largest_niche: None, align: AbiAndPrefAlign { abi: Align { pow2: 1 }, pref: Align { pow2: 3 } }, size: Size { raw: 2 } }, Layout { fields: Arbitrary { offsets: [], memory_index: [] }, variants: Single { index: 1 }, abi: Aggregate { sized: true }, largest_niche: None, align: AbiAndPrefAlign { abi: Align { pow2: 1 }, pref: Align { pow2: 3 } }, size: Size { raw: 2 } }] }, abi: Aggregate { sized: true }, largest_niche: Some(Niche { offset: Size { raw: 0 }, scalar: Scalar { value: Int(I8, false), valid_range: 0..=1 } }), align: AbiAndPrefAlign { abi: Align { pow2: 1 }, pref: Align { pow2: 3 } }, size: Size { raw: 2 } } })
`

**Comment by RalfJung** [comment]
So looks like we have that same problem both in the CTFE engine and in the codegen backend. Fun. At least we are consistent. ;)

**Comment by JohnTitor** [comment]
Fixed by #92932, closing.
